src/Libraries/Microsoft.Extensions.AI/ChatCompletion/ChatClientBuilderServiceCollectionExtensions.cs

```diff
@@ -11,37 +11,71 @@ namespace Microsoft.Extensions.DependencyInjection;
 public static class ChatClientBuilderServiceCollectionExtensions
 {
     /// <summary>Adds a chat client to the <see cref="IServiceCollection"/>.</summary>
-    /// <param name="services">The <see cref="IServiceCollection"/> to which the client should be added.</param>
-    /// <param name="clientFactory">The factory to use to construct the <see cref="IChatClient"/> instance.</param>
-    /// <returns>The <paramref name="services"/> collection.</returns>
-    /// <remarks>The client is registered as a scoped service.</remarks>
-    public static IServiceCollection AddChatClient(
-        this IServiceCollection services,
-        Func<ChatClientBuilder, IChatClient> clientFactory)
+    /// <param name="serviceCollection">The <see cref="IServiceCollection"/> to which the client should be added.</param>
+    /// <param name="innerClient">The inner <see cref="IChatClient"/> that represents the underlying backend.</param>
+    /// <returns>A <see cref="ChatClientBuilder"/> that can be used to build a pipeline around the inner client.</returns>
+    /// <remarks>The client is registered as a singleton service.</remarks>
+    public static ChatClientBuilder AddChatClient(
+        this IServiceCollection serviceCollection,
+        IChatClient innerClient)
+        => AddChatClient(serviceCollection, _ => innerClient);
+
+    /// <summary>Adds a chat client to the <see cref="IServiceCollection"/>.</summary>
+    /// <typeparam name="T">The type of the inner <see cref="IChatClient"/> that represents the underlying backend. This will be resolved from the service provider.</typeparam>
+    /// <param name="serviceCollection">The <see cref="IServiceCollection"/> to which the client should be added.</param>
+    /// <returns>A <see cref="ChatClientBuilder"/> that can be used to build a pipeline around the inner client.</returns>
+    /// <remarks>The client is registered as a singleton service.</remarks>
+    public static ChatClientBuilder AddChatClient<T>(
+        this IServiceCollection serviceCollection)
+        where T : IChatClient
+        => AddChatClient(serviceCollection, services => services.GetRequiredService<T>());
```

Comment:
What would happen if someone literally called `AddChatClient<IChatClient>`?

Reply:
That's equivalent to doing something like:

```cs
serviceCollection.AddSingleton(services => services.GetRequiredService<IMyInterface>());
```

i.e., registering some `T` service as a factory that itself tries to resolve a `T`. In practice this behaves as an infinite loop, both in the `AddChatClient<IChatClient>` case and in the line of code I posted in this comment. That's not great but it's the same as the more general behavior in the case of recursive registrations.

There is another design aspect that's a bit ambiguous here which is that if you do:

```cs
serviceCollection.AddChatClient<SomeChatClient>();
```

... or similar, then this does *not* register `SomeChatClient` in DI, but rather only registers `IChatClient`. That makes sense in the context of this scenario alone, but feels inconsistent with other DI `Add...<T>` methods such as `AddSingleton<T>`, since they actually do register the `T` service in DI.

We could fix this by calling `TryAddSingleton<T>()` inside the `AddChatClient<T>` method. However it's not clear on how often that would be useful in practice since in most cases you may need to call some other method to register the concrete `T` service anyway (e.g., because it needs some other params).

Given these two pitfalls I'm not 100% sold on adding the generic-typed overloads here. Do you have an opinion about it? I'm probably leaning towards not having the generic-typed overloads, since then we're free to add them in the future if we get enough evidence of the need.

Reply:
Update: to keep this moving forwards, I've removed them.

Reply:
Sounds good